File: align-android/app/src/main/java/ro/align/app/net/AlignApi.kt
package ro.align.app.net

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import org.webrtc.PeerConnection
import ro.align.app.BuildConfig
import java.util.concurrent.TimeUnit

/** REST Align: token semnalizare + ICE (același contract ca web). */
class AlignApi(
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()
) {
    private val base = BuildConfig.API_BASE_URL.trimEnd('/')

    suspend fun fetchSignalingToken(sessionToken: String, userId: String): String = withContext(Dispatchers.IO) {
        val req = Request.Builder()
            .url("$base/api/call/signaling-token")
            .header("x-session-token", sessionToken)
            .header("x-user-id", userId)
            .get()
            .build()
        client.newCall(req).execute().use { res ->
            val body = res.body?.string().orEmpty()
            if (!res.isSuccessful) throw ApiException("signaling-token ${res.code}: $body")
            val j = JSONObject(body)
            return@withContext j.getString("token")
        }
    }

    /**
     * Anunță pe server că sunăm pe [toId] — înregistrează pending + trimite FCM/Web Push la callee.
     * Fără asta, apelul din app nativ pornește doar WebSocket-ul și browserul nu vede „te sună”.
     */
    suspend fun ringCallee(
        sessionToken: String,
        userId: String,
        toId: String,
        roomId: String,
        audioOnly: Boolean,
    ) = withContext(Dispatchers.IO) {
        val json = JSONObject()
            .put("toId", toId)
            .put("roomId", roomId)
            .put("audioOnly", audioOnly)
            .toString()
        val body = json.toRequestBody(JSON_MEDIA)
        val req = Request.Builder()
            .url("$base/api/call/ring")
            .header("x-session-token", sessionToken)
            .header("x-user-id", userId)
            .post(body)
            .build()
        client.newCall(req).execute().use { res ->
            val resBody = res.body?.string().orEmpty()
            if (!res.isSuccessful) throw ApiException("ring ${res.code}: $resBody")
        }
    }

    suspend fun fetchIceServers(sessionToken: String, userId: String): List<PeerConnection.IceServer> =
        withContext(Dispatchers.IO) {
            val req = Request.Builder()
                .url("$base/api/call/ice-config")
                .header("x-session-token", sessionToken)
                .header("x-user-id", userId)
                .get()
                .build()
            client.newCall(req).execute().use { res ->
                val body = res.body?.string().orEmpty()
                if (!res.isSuccessful) throw ApiException("ice-config ${res.code}: $body")
                val j = JSONObject(body)
                val arr = j.getJSONArray("iceServers")
                val out = ArrayList<PeerConnection.IceServer>()
                for (i in 0 until arr.length()) {
                    val o = arr.getJSONObject(i)
                    val urlsJson = o.get("urls")
                    val urls = when (urlsJson) {
                        is String -> listOf(urlsJson)
                        is JSONArray -> List(urlsJson.length()) { urlsJson.getString(it) }
                        else -> emptyList()
                    }
                    val user = o.optString("username", "")
                    val pass = o.optString("credential", "")
                    if (user.isEmpty() || pass.isEmpty()) {
                        throw ApiException("ice-config: TURN_REQUIRED missing username/credential")
                    }
                    val relayUrls = urls.filter { u ->
                        val x = u.trim().lowercase()
                        x.startsWith("turn:") || x.startsWith("turns:")
                    }
                    if (relayUrls.isEmpty()) {
                        throw ApiException("ice-config: TURN_REQUIRED no turn:/turns: URIs in iceServers")
                    }
                    val builder = PeerConnection.IceServer.builder(relayUrls)
                    builder.setUsername(user).setPassword(pass)
                    out.add(builder.createIceServer())
                }
                if (out.isEmpty()) throw ApiException("ice-config: TURN_REQUIRED empty iceServers")
                return@withContext out
            }
        }

    fun registerFcmToken(sessionToken: String, userId: String, fcmToken: String) {
        val json = JSONObject().put("token", fcmToken).put("platform", "android").toString()
        val body = json.toRequestBody(JSON_MEDIA)
        val req = Request.Builder()
            .url("$base/api/me/push-token")
            .header("x-session-token", sessionToken)
            .header("x-user-id", userId)
            .post(body)
            .build()
        client.newCall(req).execute().close()
    }

    companion object {
        private val JSON_MEDIA = "application/json; charset=utf-8".toMediaType()
    }

    class ApiException(message: String) : Exception(message)
}

File: align-ios/Align/Network/AlignAPI.swift
import Foundation

enum AlignAPIError: Error {
    case badStatus(Int, String)
    case decode
}

enum AlignAPI {
    static func fetchSignalingToken(session: String, userId: String) async throws -> String {
        let url = URL(string: "\(AlignConfig.apiBase)/api/call/signaling-token")!
        var req = URLRequest(url: url)
        req.addValue(session, forHTTPHeaderField: "x-session-token")
        req.addValue(userId, forHTTPHeaderField: "x-user-id")
        let (data, res) = try await URLSession.shared.data(for: req)
        guard let http = res as? HTTPURLResponse else { throw AlignAPIError.decode }
        guard http.statusCode == 200 else {
            throw AlignAPIError.badStatus(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }
        let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        guard let token = obj?["token"] as? String else { throw AlignAPIError.decode }
        return token
    }

    static func fetchIceServers(session: String, userId: String) async throws -> [[String: Any]] {
        let url = URL(string: "\(AlignConfig.apiBase)/api/call/ice-config")!
        var req = URLRequest(url: url)
        req.addValue(session, forHTTPHeaderField: "x-session-token")
        req.addValue(userId, forHTTPHeaderField: "x-user-id")
        let (data, res) = try await URLSession.shared.data(for: req)
        guard let http = res as? HTTPURLResponse, http.statusCode == 200 else {
            throw AlignAPIError.badStatus((res as? HTTPURLResponse)?.statusCode ?? 0, "")
        }
        let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        guard let arr = obj?["iceServers"] as? [[String: Any]] else { throw AlignAPIError.decode }
        return arr
    }

    static func ringCallee(session: String, userId: String, toId: String, roomId: String, audioOnly: Bool) async throws {
        let url = URL(string: "\(AlignConfig.apiBase)/api/call/ring")!
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.addValue("application/json", forHTTPHeaderField: "Content-Type")
        req.addValue(session, forHTTPHeaderField: "x-session-token")
        req.addValue(userId, forHTTPHeaderField: "x-user-id")
        let body: [String: Any] = [
            "toId": toId,
            "roomId": roomId,
            "audioOnly": audioOnly,
        ]
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, res) = try await URLSession.shared.data(for: req)
        guard let http = res as? HTTPURLResponse else { throw AlignAPIError.decode }
        guard (200 ..< 300).contains(http.statusCode) else {
            throw AlignAPIError.badStatus(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }
    }

    static func registerVoipToken(session: String, userId: String, voipToken: String) async throws {
        let url = URL(string: "\(AlignConfig.apiBase)/api/me/push-token")!
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.addValue("application/json", forHTTPHeaderField: "Content-Type")
        req.addValue(session, forHTTPHeaderField: "x-session-token")
        req.addValue(userId, forHTTPHeaderField: "x-user-id")
        let body: [String: Any] = ["apnsVoipToken": voipToken, "platform": "ios"]
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (_, res) = try await URLSession.shared.data(for: req)
        guard let http = res as? HTTPURLResponse, (200 ..< 300).contains(http.statusCode) else {
            throw AlignAPIError.badStatus((res as? HTTPURLResponse)?.statusCode ?? 0, "")
        }
    }
}
